File: src/converters/converter.py
#!/usr/bin/env python3

def numeric_to_fasta(numeric_content, sequence_name="seq"):
    """Convertit une séquence numérique en format FASTA."""
    numbers = numeric_content.strip().split()
    # Mapping des nombres en lettres (0->A, 1->B, etc.)
    mapping = {str(i): chr(65 + i) for i in range(26)}
    
    sequence = ''.join(mapping[num] for num in numbers if num in mapping)
    return f">{sequence_name}\n{sequence}\n"

def fasta_to_numeric(fasta_content):
    """Convertit une séquence FASTA en format numérique."""
    lines = fasta_content.strip().split('\n')
    sequence = ''
    for line in lines:
        if not line.startswith('>'):
            sequence += line.strip()
    
    # Mapping des lettres en nombres (A->0, B->1, etc.)
    mapping = {chr(65 + i): str(i) for i in range(26)}
    numbers = [mapping[char] for char in sequence.upper() if char in mapping]
    
    return ' '.join(numbers)

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 4:
        print("Usage: converter.py [to-fasta|to-numeric] input_file output_file")
        sys.exit(1)
        
    mode = sys.argv[1]
    input_file = sys.argv[2]
    output_file = sys.argv[3]
    
    with open(input_file, 'r') as f:
        content = f.read()
    
    if mode == 'to-fasta':
        result = numeric_to_fasta(content)
    elif mode == 'to-numeric':
        result = fasta_to_numeric(content)
    else:
        print("Mode invalide. Utilisez 'to-fasta' ou 'to-numeric'")
        sys.exit(1)
        
    with open(output_file, 'w') as f:
        f.write(result) 

File: src/converters/converter.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_LINE 10000
#define MAX_SEQ 100000

void numeric_to_fasta(const char* input, char* output) {
    char* token;
    char* rest = strdup(input);
    char sequence[MAX_SEQ] = "";
    int num;
    
    while ((token = strtok_r(rest, " \n", &rest))) {
        num = atoi(token);
        if (num >= 0 && num < 26) {
            char letter = 'A' + num;
            strncat(sequence, &letter, 1);
        }
    }
    
    sprintf(output, ">seq\n%s\n", sequence);
    free(rest);
}

void fasta_to_numeric(const char* input, char* output) {
    char* line;
    char* rest = strdup(input);
    char numbers[MAX_SEQ] = "";
    char temp[10];
    
    while ((line = strtok_r(rest, "\n", &rest))) {
        if (line[0] != '>') {
            for (int i = 0; line[i]; i++) {
                if (isalpha(line[i])) {
                    int num = toupper(line[i]) - 'A';
                    sprintf(temp, "%d ", num);
                    strcat(numbers, temp);
                }
            }
        }
    }
    
    strcpy(output, numbers);
    free(rest);
}

int main(int argc, char* argv[]) {
    if (argc != 4) {
        printf("Usage: %s [to-fasta|to-numeric] input_file output_file\n", argv[0]);
        return 1;
    }
    
    char input_buffer[MAX_SEQ];
    char output_buffer[MAX_SEQ];
    FILE *fin, *fout;
    
    fin = fopen(argv[2], "r");
    if (!fin) {
        printf("Erreur: Impossible d'ouvrir le fichier d'entrée\n");
        return 1;
    }
    
    // Lire le fichier d'entrée
    size_t total = 0;
    while (fgets(input_buffer + total, MAX_SEQ - total, fin) != NULL) {
        total += strlen(input_buffer + total);
    }
    fclose(fin);
    
    // Conversion
    if (strcmp(argv[1], "to-fasta") == 0) {
        numeric_to_fasta(input_buffer, output_buffer);
    } else if (strcmp(argv[1], "to-numeric") == 0) {
        fasta_to_numeric(input_buffer, output_buffer);
    } else {
        printf("Mode invalide. Utilisez 'to-fasta' ou 'to-numeric'\n");
        return 1;
    }
    
    // Écrire le résultat
    fout = fopen(argv[3], "w");
    if (!fout) {
        printf("Erreur: Impossible d'ouvrir le fichier de sortie\n");
        return 1;
    }
    
    fprintf(fout, "%s", output_buffer);
    fclose(fout);
    
    return 0;
} 